[23.0 backport] image/cache: Ignore Build and Revision on Windows

vvoland opened the pull request:
- backport: https://github.com/moby/moby/pull/47330
- backport: https://github.com/moby/moby/pull/47342
- fixes: https://github.com/moby/moby/issues/47307

The compatibility depends on whether `hyperv` or `process` container isolation is used.
This fixes cache not being used when building images based on older Windows versions on a newer Windows host.

**- How to verify it**

**- Description for the changelog**
```release-note
- Windows: Fix cache not being used when building images based on Windows versions older than the host's version
```

**- A picture of a cute animal (not mandatory but encouraged)**

vvoland committed:
image/cache: Ignore Build and Revision on Windows

The compatibility depends on whether `hyperv` or `process` container
isolation is used.
This fixes cache not being used when building images based on older
Windows versions on a newer Windows host.

Signed-off-by: Paweł Gronowski <pawel.gronowski@docker.com>
(cherry picked from commit 91ea04089b75f3c56e60e33e27daeb0d494d79f5)
Signed-off-by: Paweł Gronowski <pawel.gronowski@docker.com>

thaJeztah reviewed (approved):
LGTM

thaJeztah commented:
I think we're missing an import in this branch as well;

```
image/cache/cache.go:1: : # github.com/docker/docker/image/cache [github.com/docker/docker/image/cache.test]
image/cache/compare_test.go:135:22: undefined: platforms.Platform
image/cache/compare_test.go:136:22: undefined: platforms.Platform
image/cache/compare_test.go:141:24: undefined: platforms.Platform
image/cache/compare_test.go:142:24: undefined: platforms.Platform
image/cache/compare_test.go:147:24: undefined: platforms.Platform
image/cache/compare_test.go:148:24: undefined: platforms.Platform
image/cache/compare_test.go:153:24: undefined: platforms.Platform
image/cache/compare_test.go:154:24: undefined: platforms.Platform
image/cache/compare_test.go:159:24: undefined: platforms.Platform
image/cache/compare_test.go:160:24: undefined: platforms.Platform
```

vvoland committed:
image/cache: Use Platform from ocispec

Signed-off-by: Paweł Gronowski <pawel.gronowski@docker.com>
(cherry picked from commit 2c01d53d9692250a7d1b18742f1e100e43615586)
Signed-off-by: Paweł Gronowski <pawel.gronowski@docker.com>

vvoland committed:
image/cache: Require Major and Minor match for Windows OSVersion

The platform comparison was backported from the branch that vendors
containerd 1.7.

In this branch the vendored containerd version is older and doesn't have
the same comparison logic for Windows specific OSVersion.

Require both major and minor components of Windows OSVersion to match.

Signed-off-by: Paweł Gronowski <pawel.gronowski@docker.com>
(cherry picked from commit b3888ed89951e568c1bf33b5223e3e515019e87f)
Signed-off-by: Paweł Gronowski <pawel.gronowski@docker.com>

thaJeztah reviewed (approved):
LGTM
